asdf/_node_info.py

```diff
@@ -20,6 +20,67 @@ def _filter_tree(info, filters):
     return len(info.children) > 0 or all(f(info.node, info.identifier) for f in filters)
 
 
+def _get_matching_schema_property(schema, key):
+    if "properties" in schema:
+        props = schema["properties"]
+        if key in props:
+            return props[key]
+        if "patternProperties" in props:
+            patterns = props["patternProperties"]
+            for regex in patterns:
+                if re.search(regex, key):
+                    return patterns[regex]
+    return None
+
+
+def _get_subschema_for_property(schema, key):
+    # This does NOT handle $ref the expectation is that the schema
+    # is loaded with resolve_references=True
+    applicable = []
+
+    # first check properties and patternProperties
+    subschema = _get_matching_schema_property(schema, key)
+    if subschema is not None:
+        applicable.append(subschema)
+
+    # next handle schema combiners
+    if "not" in schema:
+        subschema = _get_subschema_for_property(schema["not"], key)
+        if subschema is not None:
+            # We can't resolve a valid subschema under a "not" since
+            # we'd have to know how to invert a schema
+            return None
+
+    for combiner in ("allOf", "oneOf", "anyOf"):
+        for combined_schema in schema.get(combiner, []):
+            subschema = _get_subschema_for_property(combined_schema, key)
+            if subschema is not None:
+                applicable.append(subschema)
+
+    # only return the subschema if we found exactly 1 applicable
+    if len(applicable) == 1:
+        return applicable[0]
+    return None
+
+
+def _get_schema_key(schema, key):
```

Comment:
I think it would be helpful to add explanations of what distinguishes `_get_schema_key` from `_get_subschema_for_property`.  In both cases key is an argument, in the first the key presumably is expected to be one of the properties, and in the second, `key` is not a property. Examples of the different usage contexts to distinguish the two might be useful.

Reply:
Good idea! How does e39d1e18b83e3c2b3099cac7bff7a4e85f34ebc6 look?

Reply:
LGTM